### ISSUE TITLE
Enforcing dark mode not working in production

### ISSUE THREAD

**Issue by maddin79** [facebook/docusaurus#10308]
### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#issues).

### Prerequisites

- [X] I'm using the latest version of Docusaurus.
- [X] I have tried the `npm run clear` or `yarn clear` command.
- [X] I have tried `rm -rf node_modules yarn.lock package-lock.json` and re-installing packages.
- [X] I have tried creating a repro with https://new.docusaurus.io.
- [ ] I have read the console error message carefully (if applicable).

### Description

Hello everyone, 

I hope someone can point me to a solution. Searching already for hours. 

I try to load my website only in dark mode and use for that the settings:

```
colorMode: {
            defaultMode: 'dark',
            disableSwitch: false,
            respectPrefersColorScheme: false,
        },
```

It is working fine locally (using Webstorm) with `docusaurus start` and `docusaurus build & docusaurus serve`. But if I deploy the website to my NGINX server, the light mode is loading. If I enable the switch and activate the dark mode, everything works as expected. I can not find a solution to enforce the dark mode in production. 

Any help is more then welcome.

Best 
Martin 

### Steps to reproduce

1. Create new project with 'npx create-docusaurus@latest my-website classic --typescript'
2. Build the project 
3. Deploy on NGINX

### Expected behavior

Dark mode is always enabled

### Actual behavior

Website loads with light mode. 

### Your environment

- Docusaurus version used: 3.4.0
- Environment name and version (e.g. Chrome 89, Node.js 16.4): Node.js 20.11.1
- Operating system and version (e.g. Ubuntu 20.04.2 LTS): (Live Deployment)  Debian 12
- NGINX version: 1.22.1

### Self-service

- [ ] I'd be willing to fix this bug myself.

**Comment by OzakIOne** [comment]
I don't have any issues trying to reproduce your configuration locally on my macbook with docker, the website loads by default with darkmode theme enabled

**Comment by maddin79** [comment]
I tried it now with a local docker NGINX too and it worked. Could lead a setting in NGINX to this problem?

**Comment by slorber** [comment]
We don't know, what could be the problem, but it seems solved. If you (and us) can reproduce the problem deterministically, we'll reopen, otherwise we assume it was a human mistake

**Comment by maddin79** [comment]
I'm sorry but it is not solved. I still have the problem on my live system. I tried only with a local docker nginx and with this the problem was not happening.

**Comment by Josh-Cena** [comment]
Did you try loading it in private browsing? Did you try any other hosting solution than Nginx? If this is specific to one setup it doesn't sound like a Docusaurus bug and should not be reported here.

**Comment by slorber** [comment]
If we cannot see the live system, its source code, and config, we cannot help more.

It is possible that you encounter a storage key conflict, see https://docusaurus.io/blog/releases/3.4#site-storage---experimental for details

**Comment by maddin79** [comment]
I also think that it can not be Docusaurus, because locally it works. That is really weird.

Yes, I tried private browsing @Josh-Cena, with no luck. 

@slorber I'll check that with the key conflict. And I will copy my production config into a docker container to play around with it. I'll come back to you with the result. Thanks for the help.

**Comment by slorber** [comment]
The easiest for us is that you share the URL of your live deployment.

**Comment by maddin79** [comment]
@slorber Sorry, late reply. Something else popped up. 
Yes, of course. Here it is: https://md.turn-it.org/

**Comment by slorber** [comment]
Your site has CSP policies that prevent our inline script from running, you can see errors in the console logs.

> md.turn-it.org/:6 Refused to execute inline script because it violates the following Content Security Policy directive: "script-src 'self'". Either the 'unsafe-inline' keyword, a hash ('sha256-O8zYuOjyuzUZDv3fub7DKfAs5TEd1dG+fz+hCSCFmQA='), or a nonce ('nonce-...') is required to enable inline execution.


Here's what this script is doing FYI, it is required to get the theme work and avoid a flash of wrong theme before React hydrates

```js
!(function () {
  function t(t) {
    document.documentElement.setAttribute("data-theme", t);
  }
  var e =
    (function () {
      try {
        return new URLSearchParams(window.location.search).get(
          "docusaurus-theme",
        );
      } catch (t) {}
    })() ||
    (function () {
      try {
        return window.localStorage.getItem("theme");
      } catch (t) {}
    })();
  t(null !== e ? e : "dark");
})(),
  (function () {
    try {
      const n = new URLSearchParams(window.location.search).entries();
      for (var [t, e] of n)
        if (t.startsWith("docusaurus-data-")) {
          var a = t.replace("docusaurus-data-", "data-");
          document.documentElement.setAttribute(a, e);
        }
    } catch (t) {}
  })();
```

**Comment by maddin79** [comment]
Hey @slorber, thank you so much. I totally overlooked this error. It works now :) .
Are there any plans to remove the inline scripts, if that is even possible?

**Comment by slorber** [comment]
Afaik it's not really possible to remove the inline scripts. At best we could make it so that it works for your specific use-case by hardcoding the dark `data-theme` attribute directly in the HTML doc, but for all other cases we'd still need the script to run.

This has been explained a bit here: https://github.com/facebook/docusaurus/discussions/10231

**Comment by maddin79** [comment]
Ah, ok. For me it is fine and I also support your point of view [#10231](https://github.com/facebook/docusaurus/discussions/10231). I'm not an expert in XSS or general website hacking, but I also do not really see an attack surface in this kind of website.

Thanks for the fast help and Docusaurus. Really appreciate it.

**Comment by manuel-rw** [comment]
Hi, experiencing the same issue. Page is up at https://homarr.dev.
Default [should be ``dark``](https://github.com/homarr-labs/documentation/blob/dc5c897d4a3aae72be1a08b87f683ac7b6683485/docusaurus.config.ts#L230), but it always is ``light``. I see no scripts being blocked by CSP. I can open a new issue if you don't want necroposts.

**Comment by slorber** [comment]
@manuel-rw when I go to your site, this is what I see:

![Image](https://github.com/user-attachments/assets/9c2556d8-57f9-445b-8c4f-5443ab8b4a35)

Isn't this "dark"? Do you see something else?

Have you tried locally or in production?

Have you tried in another browser or a new browser guest session?

Can you check if you already have a `theme` localStorage key in devtools?

I'd recommend turning this "storage namespacing" feature so that we add a hash to all localStorage keys, this may help prevent conflicts with other things on the same domain that may use the exact same keys.

https://docusaurus.io/blog/releases/3.4#site-storage---experimental

**Comment by manuel-rw** [comment]
Thanks for the reply @slorber .
I saw something else on my browsers (2 different devices), but it seems that an extension is interfering with it.
When I tried in safe mode with no extensions enabled, I got ``dark`` as the default too...
Sorry that I didn't check that deep myself.
The ``theme`` key was not set in the storage of Firefox, both on working and non-working browsers.
I'll configure the experimental storage and you can consider this as resolved - I don't think this is a Docusaurus issue.
